Bean Param init issue -  java.lang.NoClassDefFoundError: io/quarkus/generated/int$quarkusrestparamConverter$

### Describe the bug

Application not able to start due to an error related with initialisation of Bean Param in a resource class


**Pagination Class**

```java
public class PageRequest implements Page {

  @QueryParam("page")
  @DefaultValue("1")
  int page;

  @QueryParam("pageSize")
  @DefaultValue("20")
  int pageSize;

  public PageRequest() {
  }

  PageRequest(final int page, final int pageSize) {
    this.page = page;
    this.pageSize = pageSize;
  }
...
}
```


**Usage of PageRequest class**

`Response get(@BeanParam PageRequest page, @QueryParam("type") NotificationType type);`








### Expected behavior

Able to inject bean parameter and build project

### Actual behavior

```
[ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 3, Time elapsed: 18.85 s <<< FAILURE! -- in com.app.integration.AssetsResourceIntegrationTest
[ERROR] com.app.integration.AssetsResourceIntegrationTest.testIfAbleToCreateAssets -- Time elapsed: 0.004 s <<< ERROR!
java.lang.RuntimeException: java.lang.ExceptionInInitializerError
        at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:643)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:727)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:534)
        at java.base/java.lang.Class.forName(Class.java:513)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:269)
        at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:252)
        at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:610)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:660)
        ... 1 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 8 more
Caused by: java.lang.NoClassDefFoundError: io/quarkus/generated/int$quarkusrestparamConverter$
        at com.app.pagination.PageRequest.__quarkus_init_converter__page(PageRequest.java)
        at io.quarkus.rest.runtime.__QuarkusInit.init(Unknown Source)
        at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:152)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy_17(Unknown Source)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy(Unknown Source)
```

### How to Reproduce?


**Reproduce simple way:**

1. Create Project
2. Create class type of Bean Param
3. Use that bean param in a resource  API parameters
4. Build project (mvn clean install)




**Reproduce the same way:**
Reproduce:
1. Create project
2. Create Bean Param class in a library
3. Create another project
5. Import library in that project
6. Use that bean parameter as resource parameter
7. Build project (mvn clean install)





### Output of `uname -a` or `ver`

Darwin LIS-C-000XT 23.4.0 Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)

### Quarkus version or git rev

3.9.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.4

### Additional information

- Related problem has been found here: https://github.com/quarkusio/quarkus/issues/39598
- With Quarkus Framework v3.8.2 the issues does not exist

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/scanning/ClassInjectorTransformer.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/scanning/ClassInjectorTransformer.java
@@ -36,6 +36,7 @@ import org.jboss.resteasy.reactive.server.core.parameters.converters.ArrayConver
 import org.jboss.resteasy.reactive.server.core.parameters.converters.DelegatingParameterConverterSupplier;
 import org.jboss.resteasy.reactive.server.core.parameters.converters.ParameterConverter;
 import org.jboss.resteasy.reactive.server.core.parameters.converters.ParameterConverterSupplier;
+import org.jboss.resteasy.reactive.server.core.parameters.converters.ParameterConverterSupport;
 import org.jboss.resteasy.reactive.server.core.parameters.converters.RuntimeResolvedConverter;
 import org.jboss.resteasy.reactive.server.injection.ResteasyReactiveInjectionContext;
 import org.jboss.resteasy.reactive.server.injection.ResteasyReactiveInjectionTarget;
@@ -76,6 +77,9 @@ public class ClassInjectorTransformer implements BiFunction<String, ClassVisitor
     public static final String INIT_CONVERTER_METHOD_NAME = "__quarkus_init_converter__";
     private static final String INIT_CONVERTER_FIELD_NAME = "__quarkus_converter__";
     private static final String INIT_CONVERTER_METHOD_DESCRIPTOR = "(" + QUARKUS_REST_DEPLOYMENT_DESCRIPTOR + ")V";
+
+    private static final String PARAMETER_CONVERTER_SUPPORT_BINARY_NAME = ParameterConverterSupport.class.getName().replace('.',
+            '/');
 
     private static final String MULTIPART_SUPPORT_BINARY_NAME = MultipartSupport.class.getName().replace('.', '/');
 
@@ -457,18 +461,14 @@ public class ClassInjectorTransformer implements BiFunction<String, ClassVisitor
                 initConverterMethod.visitJumpInsn(Opcodes.IFNONNULL, notNull);
                 // stack: [converter]
                 initConverterMethod.visitInsn(Opcodes.POP);
-                // stack: []
-                // let's instantiate our delegate
-                initConverterMethod.visitTypeInsn(Opcodes.NEW, delegateBinaryName);
-                // stack: [converter]
-                initConverterMethod.visitInsn(Opcodes.DUP);
-                // stack: [converter, converter]
-                initConverterMethod.visitMethodInsn(Opcodes.INVOKESPECIAL, delegateBinaryName, "<init>",
-                        "()V", false);
-                // stack: [converter]
-                // If we don't cast this to ParameterConverter, ASM in computeFrames will call getCommonSuperType
-                // and try to load our generated class before we can load it, so we insert this cast to avoid that
-                initConverterMethod.visitTypeInsn(Opcodes.CHECKCAST, PARAMETER_CONVERTER_BINARY_NAME);
+
+                // className param
+                initConverterMethod.visitLdcInsn(delegateBinaryName.replace('/', '.'));
+                initConverterMethod.visitMethodInsn(Opcodes.INVOKESTATIC, PARAMETER_CONVERTER_SUPPORT_BINARY_NAME,
+                        "create",
+                        "(" + STRING_DESCRIPTOR + ")" + PARAMETER_CONVERTER_DESCRIPTOR,
+                        false);
+
                 // end default delegate
                 initConverterMethod.visitLabel(notNull);
             }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/ParameterConverterSupport.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/converters/ParameterConverterSupport.java
@@ -1,0 +1,25 @@
+package org.jboss.resteasy.reactive.server.core.parameters.converters;
+
+/**
+ * This class isn't used directly, it is however used by generated code meant to deal with {@link ParameterConverter}.
+ */
+@SuppressWarnings("unused")
+public final class ParameterConverterSupport {
+
+    private ParameterConverterSupport() {
+    }
+
+    /**
+     * Normally the reflective instantiation would not be needed, and we could just instantiate normally,
+     * however that could break dev-mode when the converters are in a different module and non-standard Maven
+     * configuration is used (see <a href="https://github.com/quarkusio/quarkus/issues/39773#issuecomment-2030493539">this</a>)
+     */
+    public static ParameterConverter create(String className) {
+        try {
+            Class<?> clazz = Class.forName(className, true, Thread.currentThread().getContextClassLoader());
+            return (ParameterConverter) clazz.getConstructor().newInstance();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to create instance of " + className, e);
+        }
+    }
+}

